### PR TITLE
Convert a deal.II tria to a CGAL::Surface_mesh

### PR DESCRIPTION
**Pull request by fdrmrc**
Converts a Triangulation<3> to a CGAL::Surface_mesh describing its boundary by looping over the boundary faces. With this, one can then use #13717 and #13664 to create a deal.II Triangulation<2,3> out of boolean operations with other deal.II triangulations.

### CODE REVIEW

File: source/cgal/surface_mesh.cc [hunk 1]
```diff
@@ -114,10 +127,60 @@ namespace CGALWrappers
...
+    CGAL::Surface_mesh<CGALPointType> &         mesh)
+  {
+    Assert(tria.n_cells() > 0,
+           ExcMessage(
+             "Triangulation cannot be empty upon calling this function."));
+    Assert(mesh.is_empty(),
+           ExcMessage(
+             "The surface mesh must be empty upon calling this function."));
+    Assert((dynamic_cast<const MappingQ<dim, spacedim> *>(&mapping) != nullptr),
+           ExcMessage("Not yet implemented for this Mapping."));
```
Comment (luca-heltai) [inline source/cgal/surface_mesh.cc:146]: You are not using the mapping. Remove it from the input arguments, and remove this assert.

File: source/cgal/surface_mesh.cc [hunk 2]
```diff
@@ -114,10 +127,60 @@ namespace CGALWrappers
...
+          {
+            dealii_cell_to_cgal_surface_mesh(cell, mapping, mesh);
+          }
+      }
+    else if constexpr (dim == 3 && spacedim == 3)
+      {
+        for (const auto &cell : tria.active_cell_iterators())
+          {
+            const auto &face_indices = cell->face_indices();
+            for (const auto f : face_indices)
```
Comment (luca-heltai) [inline source/cgal/surface_mesh.cc:164]:
```suggestion
            for (const auto &f : cell->face_indices())
```

File: source/cgal/surface_mesh.cc [hunk 3]
```diff
@@ -114,10 +127,60 @@ namespace CGALWrappers
...
+              {
+                if (cell->face(f)->at_boundary())
```
Comment (luca-heltai) [inline source/cgal/surface_mesh.cc:166]:
```suggestion
                if (cell->face(f)->at_boundary())
```

File: source/cgal/surface_mesh.cc [hunk 4]
```diff
@@ -114,10 +127,60 @@ namespace CGALWrappers
...
+                  {
+                    map_vertices(cell->face(f), deal2cgal, mesh);
+                    add_facet(cell->face(f),
+                              deal2cgal,
+                              mesh,
+                              (f % 2 == 0 || cell->n_vertices() != 8));
+                  }
+              }
+          }
```
Comment (luca-heltai) [inline source/cgal/surface_mesh.cc:175]:
```suggestion
          }
```

File: source/cgal/surface_mesh.cc [hunk 5]
```diff
@@ -55,7 +55,7 @@ namespace
           Assert(false, ExcInternalError());
           break;
       }
-    if (clockwise_ordering == false)
+    if (clockwise_ordering == true)
       std::reverse(indices.begin(), indices.end());
```
Comment (luca-heltai) [inline source/cgal/surface_mesh.cc:59]: Can you isolate this change, and the changes in the existing test file outputs in a commit of its own?

**Review by luca-heltai — CHANGES_REQUESTED** [review]
Small changes from my side.

File: source/cgal/surface_mesh.cc (hunk 5 above)
Comment (fdrmrc) [reply source/cgal/surface_mesh.cc:59]: Done, thanks.

**Comment by luca-heltai** [comment]
By reading the tests, I realized that you call "stich_borders" on the resulting meshes. Can we do this inside the function that generates the surface mesh, rather than outside?

Also, if you need a triangulated surface somewhere, can't you just call "triangulate_surface", instead of throwing an execption?

**Comment by luca-heltai** [comment]
mmmm.... this can't be right:
```
if (dim < spacedim && all_reference_cells_are_hyper_cube() && all_reference_cells_are_simplex())
```
this is always going to evaluate to false, since the last two conditions are mutually exclusive. 

I think this check should be optional, and by default we should not run through it, unless we want to exclude a priori grids like moebius strips.

**Review by luca-heltai — APPROVED** [review]

**Pull request merged**